### DISCUSSION TITLE
编译 openwrt 的 Xray-croe 如何指定使用最新源码编译？

### DISCUSSION THREAD

**Discussion by chika0801** [XTLS/Xray-core#3158] [Q&A] [answered]
比如这个文件中 https://github.com/chika0801/helloworld/blob/test/xray-core/Makefile#L8

```
include $(TOPDIR)/rules.mk

PKG_NAME:=xray-core
PKG_VERSION:=1.8.9
PKG_RELEASE:=1

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://codeload.github.com/XTLS/Xray-core/tar.gz/v$(PKG_VERSION)?
PKG_HASH:=708cf7754c733c8eb98939495c0c2e698ca5712383b87dc25eea974a0d332721
```

是修改 PKG_VERSION:=1.8.9 和 PKG_HASH 来获取指定版本

我想实现每次编译时，下载最新的源码来编译。请问如何修改这个 Makefile 文件。

我问过GPT没解决这问题。

**Comment by KFERMercer** [comment, marked as answer]
```makefile
include $(TOPDIR)/rules.mk

PKG_NAME:=xray-core
PKG_VERSION:=snap-yy.mm.dd.hh.ss
PKG_RELEASE:=1

PKG_SOURCE_URL:=https://github.com/XTLS/Xray-core.git
PKG_SOURCE_PROTO:=git
PKG_SOURCE_SUBDIR=$(PKG_NAME)-$(PKG_VERSION)
PKG_SOURCE_VERSION:=master
PKG_MIRROR_HASH=skip

PKG_MAINTAINER:=rprx
PKG_LICENSE:=GPL-2.0
PKG_LICENSE_FILES:=LICENSE
```

https://github.com/search?q=%22PKG_SOURCE_VERSION%3A%3Dmaster%22&type=code

**Comment by chika0801** [comment]
```
include $(TOPDIR)/rules.mk

PKG_NAME:=xray-core
PKG_VERSION:=1.8.9
PKG_RELEASE:=1

PKG_SOURCE_SUBDIR=$(PKG_NAME)-$(PKG_VERSION)
PKG_SOURCE_URL:=https://github.com/XTLS/Xray-core.git
PKG_SOURCE_VERSION:=main
PKG_SOURCE_PROTO:=git
PKG_MIRROR_HASH=skip
```

Thank you for your reply. I tested this method and it failed to compile.

https://github.com/chika0801/immortalwrt/actions/runs/8337191762

**Comment by KFERMercer** [comment]
```log
2024-03-19T04:08:00.0081593Z Patch failed!  Please fix ./patches/010-go1.21.patch!
```

虽然这个错误不在本题讨论范围内，~~但是直接把[这个](https://github.com/chika0801/helloworld/blob/test/xray-core/patches/010-go1.21.patch)删了就好。~~ ←忽略

更新：

看了下你的库，解决方案其实是把[这个](https://github.com/immortalwrt/packages/tree/master/lang/golang)放进去。

上边那个patch，如果你移植了go-1.22到op上就得删，反之就不用删。

再更新：

把一楼点成 answer ，谢谢

**Comment by chika0801** [comment]
感谢你的回复。不懂op编译方面的知识，都是入门极操作见谅。

我现在测试的：

makefile的文件夹，删除了 `patches/010-go1.21.patch`
https://github.com/chika0801/helloworld/blob/test/xray-core

编译固件的文件
https://github.com/chika0801/immortalwrt/blob/test/.github/workflows/AX6000.yml

失败结果
https://github.com/chika0801/immortalwrt/actions/runs/8341228733

请问一下 `看了下你的库，解决方案其实是把[这个](https://github.com/immortalwrt/packages/tree/master/lang/golang)放进去。`
意思是不是我这样的操作
https://github.com/chika0801/immortalwrt/blob/test/.github/workflows/AX6000.yml#L43-L46

**Comment by KFERMercer** [comment]
把[这个](https://github.com/chika0801/helloworld/commit/437e5907427fd4efae58881ec4f64c316f3264c6)提交 revert 就可以编出来了。

你的编的op里没有go-1.22，所以不需要删：

> 上边那个patch，如果你移植了go-1.22到op上就得删，反之就不用删。

**Reply by chika0801** [reply]
我把 [这个](https://github.com/chika0801/helloworld/commit/437e5907427fd4efae58881ec4f64c316f3264c6)提交 恢复了，编译还是失败的。感谢你的帮助。

**Reply by KFERMercer** [reply]
看了一下最新的CI日志，问题似乎出在工作目录不是源码目录，进而没有 ./go.mod 这个文件，最终导致 patching 失败。

把[这行](https://github.com/chika0801/immortalwrt/blob/8e2d3925edf7762de8fc844ba7f9e940cc27b0a6/.github/workflows/AX6000.yml#L86)改为：
```shell
        make -j$(nproc) || make -j1 V=s || ls -R /home/runner/work/immortalwrt/immortalwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/Xray-core-1.8.9/
```

即可定位问题的根源。

但是我看到你好像已经弃疗了，所以要不要解决随你决定吧

**Reply by chika0801** [reply]
感谢你的回复，我一直想解决这个疑问（方便用最新源代码编译，测试一些功能）。

**Comment by chika0801** [comment]
> 把[这行](https://github.com/chika0801/immortalwrt/blob/8e2d3925edf7762de8fc844ba7f9e940cc27b0a6/.github/workflows/AX6000.yml#L86)改为：
> 
> ```shell
>         make -j$(nproc) || make -j1 V=s || ls -R /home/runner/work/immortalwrt/immortalwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/Xray-core-1.8.9/
> ```
> 
> 即可定位问题的根源。

这是改后，编译的结果。没有报错。
https://github.com/chika0801/immortalwrt/actions/runs/8346973579

**Comment by chika0801** [comment]
再次感谢你的回复，虽然我没成功。

我找到了解决方法，我去 /usr/bin 目录下，替换xray文件就能测试了。

我的路由器是红米AX6000，使用 [Xray-linux-arm64-v8a](https://github.com/XTLS/Xray-core/actions/runs/8363049370/artifacts/1343442984) 文件
